include/nbl/asset/filters/kernels/CConvolutionWeightFunction.h

```diff
@@ -0,0 +1,175 @@
+// `derivative` is a class template member not a function template member (done to handle chain rule) of CWeightFunction1D.
+// So, we would have to create new `CWeightFunction1D`s entirely and use them (not just here, but in the various specializations
+// of this function). Not sure if its worth it, both in terms of performance and code complexity.
+
+// this is the horribly slow generic version that you should not use (only use the specializations or when one of the weights is a dirac)
+template<KernelWeightFunction1D WeightFunction1DA, KernelWeightFunction1D WeightFunction1DB>
+class CConvolutionWeightFunction1D
+{
+	static_assert(std::is_same_v<impl::weight_function_value_type_t<WeightFunction1DA>, impl::weight_function_value_type_t<WeightFunction1DB>>, "Both functions must use the same Value Type!");
+
+public:
+	constexpr static inline uint32_t k_smoothness = WeightFunction1DA::k_smoothness + WeightFunction1DB::k_smoothness;
+
+	inline CConvolutionWeightFunction1D(WeightFunction1DA&& funcA, WeightFunction1DB&& funcB)
+		: m_funcA(std::move(funcA)), m_funcB(std::move(funcB)),
+		m_isFuncAWider((m_funcA.getMaxSupport() - m_funcA.getMinSupport()) > (m_funcB.getMaxSupport() - m_funcB.getMinSupport()))
+	{
+		m_minSupport = m_funcA.getMinSupport() + m_funcB.getMinSupport();
+		m_maxSupport = m_funcA.getMaxSupport() + m_funcB.getMaxSupport();
+	}
+
+	inline void stretch(const float s)
+	{
+		assert(s != 0.f);
+
+		m_minSupport *= s;
+		m_maxSupport *= s;
+
+		const float rcp_s = 1.f / s;
+
+		m_invStretch *= rcp_s;
+	}
+
+	inline void scale(const float s)
+	{
+		assert(s != 0.f);
+        m_totalScale *= s;
+    }
+
+	inline void stretchAndScale(const float stretchFactor)
+	{
+		stretch(stretchFactor);
+		scale(1.f / stretchFactor);
+	}
+
+	inline double weight(const float x, const uint8_t channel, const uint32_t sampleCount = 64u) const
+	{
+		return static_cast<double>(m_totalScale * weight_impl(x * m_invStretch, channel, sampleCount));
+	}
+
+	inline float getMinSupport() const { return m_minSupport; }
+	inline float getMaxSupport() const { return m_maxSupport; }
+
+private:
+	const WeightFunction1DA m_funcA;
+	const WeightFunction1DB m_funcB;
+
+	const bool m_isFuncAWider;
+	float m_minSupport;
+	float m_maxSupport;
+	float m_invStretch = 1.f;
+	float m_totalScale = 1.f;
+
+	double weight_impl(const float x, const uint32_t channel, const uint32_t sampleCount) const
+	{
+		if constexpr (std::is_same_v<WeightFunction1DB, CWeightFunction1D<SDiracFunction>>)
+		{
+			return m_funcA.weight(x, channel);
+		}
+		else if (std::is_same_v<WeightFunction1DA, CWeightFunction1D<SDiracFunction>>)
+		{
+			return m_funcB.weight(x, channel);
+		}
+		else
+		{
+			auto [minIntegrationLimit, maxIntegrationLimit] = getIntegrationDomain(x);
+
+			// TODO(achal): what ????
+			// if this happens, it means that `m_ratio=INF` and it degenerated into a dirac delta
+			if (minIntegrationLimit == maxIntegrationLimit)
```

Comment:
This was a comment for the original case of convolving `f*g` where `f` was supposed to be unscaled and only `g` was scaled by `m_ratio` .

Now obviously everything changed, and you take two full functions with independent scales (great design choice and simplification of the API btw)

So this was supposed to be in the case of when `m_funcB.m_invStretch > FLT_MAX` or `m_funcA.m_invStretch < FLT_MIN` (or some other epsilon)

This makes the function B degenerate to a single point (dirac delta with multiplier equal to total energy under the curve) or function A  has stretched to what is basically flat line with value `f_a(0)`

Reply:
@achalpandeyy mark as resolved when read